Implement -filter for features(-type=>'coverage')

jmarshall opened the pull request:
At present, using `features(-type=>'coverage', -filter => sub { … })` ends up with your filter subroutine being silently ignored. This fixes that, implementing it as a filter that is called once for each **read** that (now optionally) goes into the pileup from which coverage is calculated.

Also some minor warning and leak fixes. This took rather longer than I anticipated as it took a long time to figure out what the `bam_dup1()` in `hts_fetch_fun()` was all about, so I have expanded on the comments there to hopefully clarify this.

jmarshall committed:
Minor warning, memory leak, and typo fixes

Return something from bam_pileup_f function (hts_plbuf_push() ignores
the return value anyway); don't leak alignment from hts_read1() at EOF;
Newxz() writes to its first argument; reverse mapping is via a different
constant.

jmarshall committed:
Add details to hts_fetch_fun() bam_dup1() comment

In adding filtering to bami_coverage(), I originally planned to have
hts_fetch() pass both b/b_sv (a bam1_t and an SV referencing it) to
its callback functions, reusing the same C object and Perl reference
for each iteration, and not needing bam_dup1() in the C callbacks.
However this broke existing callback subroutines that expected to stash
the references without deep-copying them -- thus the real reason for
this bam_dup1().

It might still be somewhat preferable to have hts_fetch manage
both b and b_sv itself, allocating a new one (via bam_init1() and
sv_setref_pv(newSV())) for each iteration. However any gain only
barely outweighs the effort of altering the existing working code.

jmarshall committed:
Implement -filter for features(-type=>'coverage') / bami_coverage()

In bami_coverage(), use an iterator-based loop rather than hts_fetch()
and callbacks, as there are now several items to pass down to a callback.
As hts_plbuf_push() eventually pushes a copy of the bam1_t, the loop can
reuse a single bam1_t and a single Perl reference to it without concern.

Note that the items being filtered are *reads*, rather than pileup
positions or similar.

avullo commented:
Interestingly, travis is happy will all tested combinations of perl and htslib except 5.26/1.5 where the cram tests fail. Could you try reproducing with the same setting?

avullo commented:
For completeness, would it be possible to add similar tests in the cram/sam case? It's fine if it's not.

jmarshall commented:
Will have a look at both of these.

avullo commented:
Marvelous, thanks :+1:

jmarshall committed:
Add CRAM coverage -filter tests

Add tests similar to those added to t/01bam.t.
No SAM tests added, as coverage requires -seq_id, which requires an
index, and SAM files have no indices.

jmarshall commented:
I've added CRAM tests, but coverage is not applicable for SAM as the required `-seq_id` option requires an index.

Nothing has been changed (only a few tests added), yet travis is now happy even on the Perl 5.26/htslib 1.5 combination.

Running the _t/03cram.t_ tests under valgrind shows a double-free bug (the cram index is freed via both `hts_close` and `hts_idx_destroy`), present also on the **master** branch and with any version of HTSlib, that occasionally leads to a segfault (only once in my attempts to reproduce this). I suspect it is this that caused the failed travis run, which ran the _t/03cram.t_ tests twice: once successfully (via `./Build test`) and once with a segfault (via `travisci/harness.sh`).

avullo commented:
That's excellent @jmarshall :+1: 
To be honest, I didn't and still don't understand the reason why two sets of tests should be executed on travis. At the same time, this has proven to be useful since it's exposed the bug.
Your observation above reminds me of some additional tests I had thought to add when initially looking at the test suite, namely using Test::LeakTrace which I suspect acts on behalf of valgrind.

Thanks again for you contribution, this is good to go for me.
